Content.Server/Radio/EntitySystems/MessagesServerSystem.cs

```diff
@@ -0,0 +1,166 @@
+            }
+
+            //If the cart reports a changed name, it adds it to the toUpdate list and updates the cart's name dictionary
+            if (component.NameDict.TryGetValue(cartComponent.UserUid.Value, out var cartUserName) && cartUserName == cartComponent.UserName)
+                continue;
+
+            component.NameDict[cartComponent.UserUid.Value] = cartComponent.UserName;
+            toUpdate.Add((cartComponent.UserUid.Value, cartComponent.UserName));
+            foreach (var entry in component.NameDict)
+            {
+                if (!cartComponent.NameDict.TryGetValue(entry.Key, out var targetUserName) || targetUserName != entry.Value)
+                {
+                    cartComponent.NameDict[entry.Key] = entry.Value;
+                }
+            }
+            _messagesCartridgeSystem.ForceUpdate(cartUid, cartComponent);
+        }
+
+        //If any names were changed or added, the server updates all the carts on its map.
+        if (toUpdate.Count > 0)
+        {
+            query = _entManager.AllEntityQueryEnumerator<MessagesCartridgeComponent>();
+            while (query.MoveNext(out var cartUid, out var cartComponent))
+            {
+                if (Transform(cartUid).MapID != mapId)
+                    continue;
+                if ((cartComponent.UserUid == null || cartComponent.UserName == null) && !(_messagesCartridgeSystem.UpdateName(cartUid, cartComponent)))
+                    continue;
+                if (cartComponent.EncryptionKey != component.EncryptionKey)
+                    continue;
+
+                foreach (var (key, value) in toUpdate)
+                {
+                    cartComponent.NameDict[key] = value;
+                }
+                _messagesCartridgeSystem.ForceUpdate(cartUid, cartComponent);
+            }
+        }
+    }
+
+    //Sync function that updates the name dictionaries of all carts to match the server.
+    //Called periodically with update.
+    public void Sync(EntityUid uid, MessagesServerComponent component)
+    {
+        var mapId = Transform(uid).MapID;
+        var query = _entManager.AllEntityQueryEnumerator<MessagesCartridgeComponent>();
+        while (query.MoveNext(out var cartUid, out var cartComponent))
+        {
+            if (Transform(cartUid).MapID != mapId)
+                continue;
+            if (cartComponent.UserUid == null || cartComponent.UserName == null)
+                continue;
+            if (cartComponent.EncryptionKey != component.EncryptionKey)
+                continue;
+
+            foreach (var entry in component.NameDict)
+            {
+                if (cartComponent.NameDict.ContainsKey(entry.Key) || cartComponent.NameDict[entry.Key] != entry.Value)
+                {
+                    cartComponent.NameDict[entry.Key] = entry.Value;
+                }
+            }
+        }
+    }
+
+    //function that tries to send a message to any matching cartridges on its map
```

Comment:
I don't really get why we even need 2 steps? I get one only applies if it has power but just seems like unnecessary iteration when we could just wrap it into the 1 update loop and save a lot of performance + headaches.